File: haskell/space-age/src/SpaceAge.hs
module SpaceAge (Planet (..), ageOn) where

data Planet
  = Mercury
  | Venus
  | Earth
  | Mars
  | Jupiter
  | Saturn
  | Uranus
  | Neptune

ageOn :: Planet -> Float -> Float
ageOn planet seconds = secondsInEarthYears / planetToEarthYearRatio planet
  where
    oneEarthYearInSeconds = 31557600
    secondsInEarthYears = seconds / oneEarthYearInSeconds
    planetToEarthYearRatio :: Planet -> Float
    planetToEarthYearRatio Mercury = 0.2408467
    planetToEarthYearRatio Venus = 0.61519726
    planetToEarthYearRatio Earth = 1
    planetToEarthYearRatio Mars = 1.8808158
    planetToEarthYearRatio Jupiter = 11.862615
    planetToEarthYearRatio Saturn = 29.447498
    planetToEarthYearRatio Uranus = 84.016846
    planetToEarthYearRatio Neptune = 164.79132

File: python/space-age/space_age.py
"""Convert seconds to years on different planets in the solar system."""

from typing import Literal

Planet = Literal[
    "earth", "mercury", "venus", "mars", "jupiter", "saturn", "uranus", "neptune"
]


SECONDS_IN_PLANET_YEAR: dict[Planet, float] = {
    "earth": 31557600,
    "mercury": 7600543.81992,
    "venus": 19414149.052176,
    "mars": 59354032.69008,
    "jupiter": 374355659.124,
    "saturn": 929292362.8848,
    "uranus": 2651370019.3296,
    "neptune": 5200418560.032001,
}


class SpaceAge:
    """Converts seconds to years on different planets in the solar system."""

    def __init__(self, seconds: int):
        self.seconds = seconds

    def _on_planet(self, planet: Planet) -> float:
        return round(self.seconds / SECONDS_IN_PLANET_YEAR[planet], 2)

    def on_earth(self) -> float:
        """Return the age in Earth years"""
        return self._on_planet("earth")

    def on_mercury(self) -> float:
        """Return the age in Mercurary years"""
        return self._on_planet("mercury")

    def on_venus(self) -> float:
        """Return the age in Venus years"""
        return self._on_planet("venus")

    def on_mars(self) -> float:
        """Return the age in Mars years"""
        return self._on_planet("mars")

    def on_jupiter(self) -> float:
        """Return the age in Jupiter years"""
        return self._on_planet("jupiter")

    def on_saturn(self) -> float:
        """Return the age in Saturn years"""
        return self._on_planet("saturn")

    def on_uranus(self) -> float:
        """Return the age in Uranus years"""
        return self._on_planet("uranus")

    def on_neptune(self) -> float:
        """Return the age in Neptune years"""
        return self._on_planet("neptune")
